src/vm/callsiteinspect.cpp

```diff
@@ -0,0 +1,517 @@
+            THROWS;
+            GC_TRIGGERS;
+            MODE_ANY;
+            PRECONDITION(CheckPointer(frame));
+        }
+        CONTRACT_END;
+
+        ArgDetails details{};
+        details.Offset = pArgIter.GetNextOffset();
+        details.ElementType = pArgIter.GetArgType();
+
+#ifdef COM_STUBS_SEPARATE_FP_LOCATIONS
+        // BUGBUG do we need to handle this?
+        if ((ELEMENT_TYPE_R4 == details.ElementType || ELEMENT_TYPE_R8 == details.ElementType)
+            && TransitionBlock::IsArgumentRegisterOffset(details.Offset))
+        {
+            int iFPArg = TransitionBlock::GetArgumentIndexFromOffset(details.Offset);
+            details.Offset = static_cast<int>(frame->GetFPArgOffset(iFPArg));
+        }
+#endif // COM_STUBS_SEPARATE_FP_LOCATIONS
+
+        // Get the TypeHandle for the argument's type.
+        MetaSig *pSig = pArgIter.GetSig();
+        details.Type = pSig->GetLastTypeHandleThrowing();
+
+        if (details.ElementType == ELEMENT_TYPE_BYREF)
+        {
+            details.IsByRef = TRUE;
+            // If this is a by-ref arg, GetOBJECTREFFromStack() will dereference "addr" to
+            // get the real argument address. Dereferencing now will open a gc hole if "addr"
+            // points into the gc heap, and we trigger gc between here and the point where
+            // we return the arguments.
+
+            TypeHandle tycopy;
+            details.ElementType = pSig->GetByRefType(&tycopy);
+            if (details.ElementType == ELEMENT_TYPE_VALUETYPE)
+                details.Type = tycopy;
+        }
+#ifdef ENREGISTERED_PARAMTYPE_MAXSIZE
+        else if (details.ElementType == ELEMENT_TYPE_VALUETYPE)
+        {
+            details.IsByRef = ArgIterator::IsArgPassedByRef(details.Type);
+        }
+#endif // ENREGISTERED_PARAMTYPE_MAXSIZE
+
+        RETURN (details);
+    }
+
+    INT64 CopyOBJECTREFToStack(
+        _In_ OBJECTREF src,
+        _In_opt_ PVOID pvDest,
+        _In_ CorElementType typ,
+        _In_ TypeHandle ty,
+        _In_ MetaSig *pSig,
+        _In_ BOOL fCopyClassContents)
+    {
+        // Use local to ensure proper alignment
+        INT64 ret = 0;
+
+        CONTRACT(INT64)
+        {
+            THROWS;
+            GC_TRIGGERS;
+            MODE_COOPERATIVE;
+            INJECT_FAULT(COMPlusThrowOM());
+            PRECONDITION(CheckPointer(pvDest, NULL_OK));
+            PRECONDITION(CheckPointer(pSig));
+            PRECONDITION(typ != ELEMENT_TYPE_VOID);
+        }
+        CONTRACT_END;
+
+        if (fCopyClassContents)
+        {
+            // We have to copy the contents of a value class to pvDest
+
+            // write unboxed version back to memory provided by the caller
+            if (pvDest)
+            {
+                if (ty.IsNull())
+                    ty = pSig->GetRetTypeHandleThrowing();
+
+                _ASSERTE(src != NULL || Nullable::IsNullableType(ty));
+                ty.GetMethodTable()->UnBoxIntoUnchecked(pvDest, src);
```

Comment:
This isn't safe for HVAs on ARM64 or for complex structures on Unix ABI amd64. See various somewhat confusing uses of the ArgDestination structure for techniques on how to handle that.

Reply:
For right now this is only for Windows so I will `#define` in an error code that will be returned if this is ran on non-Windows.